### ISSUE TITLE
RRef Leak and pickler issues with Distributed Optimizer

### ISSUE THREAD

**Issue by pritamdamania87** [pytorch/pytorch#30022]
## 🐛 Bug

Script shared below runs into RRef leaks and pickler issues when running the DistributedOptimizer

## To Reproduce

```
import torch
from torch.distributed import rpc
import torch.distributed.autograd as dist_autograd
from torch import optim
from torch.distributed.optim import DistributedOptimizer
from tempfile import NamedTemporaryFile
import multiprocessing as mp

def random_tensor():
    return torch.rand((3, 3), requires_grad=True)

def _run_process(self_rank, dst_rank, file_name):
    self_name = "worker{}".format(self_rank)
    dst_name = "worker{}".format(dst_rank)

    rpc.init_model_parallel(
        self_name=self_name,
        self_rank=self_rank,
        worker_name_to_id={"worker0": 0, "worker1": 1},
        init_method="file://{}".format(file_name),
    )

    with dist_autograd.context() as context_id:
       # Forward pass.
       rref1 = rpc.remote(dst_name, random_tensor)
       rref2 = rpc.remote(dst_name, random_tensor)
       loss = rref1.to_here() + rref2.to_here()

       # Backward pass.
       dist_autograd.backward([loss.sum()])

       # Optimizer.
       try:
        dist_optim = DistributedOptimizer(
            optim.SGD,
            [rref1, rref2],
            lr=0.05,
        )
       except Exception as e:
           print (e)
           raise e
       dist_optim.step()

    rpc.join_rpc()

file_name = NamedTemporaryFile().name
processes = []
for i in range(2):
    p = mp.Process(target=_run_process, args=(i, (i + 1) % 2, file_name))
    p.start()

for p in processes:
    p.join()
```

The code above fails with: https://gist.github.com/pritamdamania87/08cf8de0aa95de03ba1e50da89f96996

It looks like when we create RRefs in the DistributedOptimizer constructor, we don't "wait" for these RRefs. Aside from the issue above, we should ensure we wait for the rrefs so that we discover any errors earlier. But now, if I change the DistributedOptimizer code to wait for rrefs, I run into the following error with the same script above:

https://gist.github.com/pritamdamania87/e637f5f8a5fede8a8102c584191c28b3

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @xush6528

**Comment by aazzolini** [comment]
There's 2 things:
1) Waiting on RRef triggers a copy to local -- which is not the intended behavior. What we want is to use rpc instead and have remote return a RRef. I will send a fix for that soon.
2) If the above is not intuitive then we should think of a better API for this functionality. It would be for example a wrapper around rpc.
3) In the long run, it should be fine to not wait on RRefs as well and the termination algorithm with take care of waiting for those?

**Comment by mrshenli** [comment]
For RRef leak, I think that's because `rref1` and `rref2` are not dereferenced by Python when you call `rpc.join_rpc()`.

**Comment by pritamdamania87** [comment]
> 1. Waiting on RRef triggers a copy to local -- which is not the intended behavior. What we want is to use rpc instead and have remote return a RRef. I will send a fix for that soon.
I was referring to the optimizer rrefs here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/optim/optimizer.py#L73. I feel we should wait for these, otherwise the error reporting to the user would be very confusing. If any of those `rpc.remote` calls fail, it will show up as an error during `optimizer.step` which would be confusing for users. Also, since the constructor of the optimizer is typically called only once, it should be fine to have an expensive wait in the beginning?

**Comment by aazzolini** [comment]
We don't have wait on RRefs. So instead of calling remote that returns an RRef and wait on a remote RRef, I'll call rpc_async that returns future<RRef> and then wait on this future. I'll send a fix soon.

**Comment by pritamdamania87** [comment]
Looks like the pickler issue is due to the fact that I was calling `rref.to_here()` and the lock object couldn't be pickled.

**Comment by aazzolini** [comment]
Just not sure how to unit test this since the unit tests seem to be passing. Is there anything in the code above that differs from the unit test except the way we terminate? (vs. in the unit tests we're adding a barrier at the end?)

**Comment by aazzolini** [comment]
@Pritam i think the leak is not really caused by what you mentioned in the issue .

Instead i was able to fix the issue by moving application code into a separate function. The leaks were real: we were callling rpc_join before deleting the objects containing RRefs.

So we need to make it clear to users that all the RRefs need to be gone before calling rpc_join. A good way of ensuring this is make sure the user/app code is enclosed by a function so that python can trigger the gc before calling rpc_join.

I'll still send a PR to use rpc_async instead, so that we can catch exceptions at construction time. But I don't think it needs to block 1.4 release.


`import torch
from torch.distributed import rpc
import torch.distributed.autograd as dist_autograd
from torch import optim
from torch.distributed.optim import DistributedOptimizer
from tempfile import NamedTemporaryFile
import multiprocessing as mp

def random_tensor():
    return torch.rand((3, 3), requires_grad=True)


def run_stuff(dst_name):
    with dist_autograd.context() as context_id:
       # Forward pass.
       rref1 = rpc.remote(dst_name, random_tensor)
       rref2 = rpc.remote(dst_name, random_tensor)
       loss = rref1.to_here() + rref2.to_here()

       # Backward pass.
       dist_autograd.backward([loss.sum()])

       # Optimizer.
       try:
        dist_optim = DistributedOptimizer(
            optim.SGD,
            [rref1, rref2],
            lr=0.05,
        )
       except Exception as e:
           print (e)
           raise e
       dist_optim.step()

    print('===  A ===')
    del dist_optim 



def _run_process(self_rank, dst_rank, file_name):
    self_name = "worker{}".format(self_rank)
    dst_name = "worker{}".format(dst_rank)

    rpc.init_rpc(
        self_name=self_name,
        self_rank=self_rank,
        worker_name_to_id={"worker0": 0, "worker1": 1},
        init_method="file://{}".format(file_name),
    )

    # only create RRefs inside of a function, not here at the top level
    run_stuff(dst_name)

    rpc.join_rpc()

file_name = NamedTemporaryFile().name
processes = []
for i in range(2):
    p = mp.Process(target=_run_process, args=(i, (i + 1) % 2, file_name))
    p.start()

for p in processes:
    p.join()
`

**Comment by aazzolini** [comment]
https://github.com/pytorch/pytorch/pull/30062

**Comment by pietern** [comment]
@pritamdamania87 You can make your launcher a bit simpler by changing:

```python
processes = []
for i in range(2):
    p = mp.Process(target=_run_process, args=(i, (i + 1) % 2, file_name))
    p.start()

for p in processes:
    p.join()
```

to:

```python
import torch.multiprocessing as mp
nprocs = 2
mp.spawn(_run_process, args=(nprocs, file_name), nprocs=nprocs)
```

The function takes `(rank, *args)` as argument, so in this case `(rank, nprocs, file_name)`. 

Also see https://pytorch.org/docs/stable/multiprocessing.html#torch.multiprocessing.spawn

**Comment by pietern** [comment]
This is related to / a duplicate of #27928.

cc @rohan-varma

**Comment by rohan-varma** [comment]
There is another simple repro in the above issue that @pietern mentioned. The RRefs leak in a seemingly simple script, so maybe we should add a warning to users mentioning they should enclose RRefs in another function until we figure out a fix?

**Comment by pritamdamania87** [comment]
I think we can close this out since we're tracking the larger RRef leak issue here: https://github.com/pytorch/pytorch/issues/31325
